js/state/syncUtil.js

```diff
@@ -280,12 +300,37 @@ module.exports.newObjectId = (objectPath) => {
 }
 
 /**
+ * Given a bookmark folder's folderId, get or set its object ID.
+ * @param {number} folderId
+ * @returns {Array.<number>}
+ */
+module.exports.findOrCreateFolderObjectId = (folderId) => {
+  if (!folderId) { return undefined }
+  const AppStore = require('../stores/appStore')
+  const appState = AppStore.getState()
+  const folderEntry = appState.get('sites').findEntry((site, _index) => {
+    return site.get('folderId') === folderId
+  })
+  if (!folderEntry) { return undefined }
+  const folderIndex = folderEntry[0]
+  const folder = folderEntry[1]
+  const objectId = folder.get('objectId')
+  if (objectId) {
+    return objectId.toJS()
+  } else {
+    return module.exports.newObjectId(['sites', folderIndex])
+  }
+}
+
+/**
  * Converts a site object into input for sendSyncRecords
  * @param {Object} site
  * @param {number=} siteIndex
+ * @param {boolean} generateObjectIds objectIds are generated via AppActions IPC
+ *   which is async. You may prefer to generate it for immediate use.
```

Comment:
so i had this issue on initial bookmarks sync because although `newObjectId()` returns the objectId, `appState.getIn(['sites', 123])` doesn't immediately reflect the objectId.
I decided to move the generation to [`sync.js`](https://github.com/brave/browser-laptop/pull/6832/files#diff-b7794baa4e0cd1fb8cab9449660662baR183) and memoize them as they're generated.